src/Quidjibo.DataProtection/Protectors/AesPayloadProtector.cs

```diff
@@ -36,65 +42,84 @@ public async Task<byte[]> ProtectAsync(byte[] payload, CancellationToken cancell
                 await cryptoStream.WriteAsync(payload, 0, payload.Length, cancellationToken);
                 cryptoStream.FlushFinalBlock();
                 var encryptedPayload = stream.ToArray();
-                var mac = ComputeMac(keys.MacKey, encryptedPayload, 0, encryptedPayload.Length);
-                var outputBuffer = new byte[encryptedPayload.Length + aes.IV.Length + mac.Length];
-                Buffer.BlockCopy(aes.IV, 0, outputBuffer, 0, aes.IV.Length);
-                Buffer.BlockCopy(mac, 0, outputBuffer, aes.IV.Length, mac.Length);
-                Buffer.BlockCopy(encryptedPayload, 0, outputBuffer, aes.IV.Length + mac.Length, encryptedPayload.Length);
+                var ivSegment = new ArraySegment<byte>(aes.IV);
+                var payloadSegment = new ArraySegment<byte>(encryptedPayload);
+                var mac = ComputeMac(keys.MacKey, version, ivSegment, payloadSegment);
+                var outputBuffer = new byte[VERSION_SIZE + IV_SIZE + MAC_SIZE + encryptedPayload.Length];
+                Memory<byte> outputBufferMemory = outputBuffer;
+                _version1.CopyTo(outputBufferMemory);
+                aes.IV.CopyTo(outputBufferMemory.Slice(VERSION_SIZE));
+                mac.CopyTo(outputBufferMemory.Slice(VERSION_SIZE + IV_SIZE));
+                encryptedPayload.CopyTo(outputBufferMemory.Slice(VERSION_SIZE + IV_SIZE + MAC_SIZE));
                 return outputBuffer;
             }
         }
 
         public async Task<byte[]> UnprotectAsync(byte[] payload, CancellationToken cancellationToken)
         {
+            if (!payload.AsSpan().Slice(0, 1).SequenceEqual(_version1))
+            {
+                throw new InvalidOperationException("Unknown payload version.");
+            }
             var keys = await ExpandKeysAsync(cancellationToken);
-            var iv = new byte[128 / 8];
-            Buffer.BlockCopy(payload, 0, iv, 0, iv.Length);
-            var mac = new byte[32];
-            Buffer.BlockCopy(payload, iv.Length, mac, 0, 32);
+            var iv = new byte[IV_SIZE];
+            Buffer.BlockCopy(payload, VERSION_SIZE, iv, 0, iv.Length);
+            var mac = new byte[MAC_SIZE];
+            Buffer.BlockCopy(payload, VERSION_SIZE + IV_SIZE, mac, 0, mac.Length);
+            VerifyMac(payload, mac, keys.MacKey);
```

Comment:
Must verify the MAC before decryption!

Reply:
I think I did that on purpose. I was thinking along on line of avoiding potential timing attacks. In retrospect, that didn't really make any sense.